File: 03-spring-examples/apps/utils/src/main/java/com/nalepa/demo/common/monitored/ExecutorsShutdownManager.java
package com.nalepa.demo.common.monitored;

import org.springframework.context.SmartLifecycle;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.*;

@Component
public class ExecutorsShutdownManager implements SmartLifecycle {

    private final List<ExecutorService> createdExecutors =
            Collections.synchronizedList(new ArrayList<>());

    private volatile boolean running = true;

    public void addExecutorService(ExecutorService executorService) {
        createdExecutors.add(executorService);
    }

    @Override
    public void start() {
    }

    @Override
    public void stop() {
        shutdownExecutors();
        running = false;
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    public void shutdownExecutors() {
        // 1. signal to end executors to end
        createdExecutors.forEach(ExecutorService::shutdown);

        // 2. parallel wait
        List<CompletableFuture<Map.Entry<Boolean, ExecutorService>>> futures = createdExecutors.stream()
                .map(executor -> CompletableFuture.supplyAsync(() -> {
                    try {
                        boolean terminated = executor.awaitTermination(5, TimeUnit.SECONDS);
                        return Map.entry(terminated, executor);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        return Map.entry(false, executor);
                    }
                }))
                .toList();

        // 3. check if executors are finished or force shutdown
        futures.forEach(future -> {
            try {
                var entry = future.get();
                if (!entry.getKey()) {
                    entry.getValue().shutdownNow();
                }
            } catch (InterruptedException | ExecutionException e) {
                Thread.currentThread().interrupt();
            }
        });
    }
}

File: spring-examples/apps/utils/src/main/kotlin/com/nalepa/demo/common/monitored/ExecutorsShutdownManager.kt
package com.nalepa.demo.common.monitored

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.binder.jvm.ExecutorServiceMetrics
import org.springframework.context.SmartLifecycle
import org.springframework.scheduling.concurrent.CustomizableThreadFactory
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.Collections
import java.util.concurrent.*


@Component
class ExecutorsShutdownManager: SmartLifecycle {

    private val createdExecutors =
        Collections.synchronizedList(mutableListOf<ExecutorService>())

    fun addExecutorService(executorService: ExecutorService) {
        createdExecutors.add(executorService)
    }


    private var running = true

    override fun start() {
    }

    override fun stop() {
        shutdownExecutors()
        running = false
    }

    override fun isRunning(): Boolean = running

    fun shutdownExecutors() {
        // 1. signal to end executors to end
        createdExecutors.forEach { it.shutdown() }

        // 2. parallel wait
        val futures = createdExecutors.map { executor ->
            CompletableFuture.supplyAsync {
                executor.awaitTermination(5, TimeUnit.SECONDS) to executor
            }
        }

        // 3. check if executors are finished or force shutdown
        futures.forEach { future ->
            val (isTerminated, executor) = future.get()
            if (!isTerminated) {
                executor.shutdownNow()
            }
        }
    }

}
